File: next/kernel/registry.py
"""Connection identity registry and plugin lifecycle publication."""

from __future__ import annotations

import time
from typing import Any

from .broker import Broker


class ConnectionRegistry:
    def __init__(self, broker: Broker) -> None:
        self.broker = broker
        self._entries: dict[str, dict[str, Any]] = {}

    @property
    def entries(self) -> list[dict[str, Any]]:
        return list(self._entries.values())

    async def register(self, hello: dict[str, Any]) -> dict[str, Any]:
        now = int(time.time() * 1000)
        manifest = hello["manifest"]
        entry = {
            "id": manifest["id"],
            "instance_id": hello.get("instance_id"),
            "manifest": manifest,
            "managed": hello["managed"],
            "conn": hello["conn"],
            "connected_at": now,
        }
        self._entries[hello["conn"]] = entry
        await self._publish_list(now)
        await self._publish_lifecycle(entry, "activated", now)
        return entry

    async def deregister(self, conn: str) -> None:
        entry = self._entries.pop(conn, None)
        if entry is None:
            return
        now = int(time.time() * 1000)
        await self._publish_list(now)
        await self._publish_lifecycle(entry, "deactivated", now)

    async def _publish_list(self, now: int) -> None:
        await self.broker.publish(
            {
                "type": "set",
                "channel": "plugins:_:list",
                "value": self.entries,
                "ts": now,
                "origin": {"plugin": "kernel", "instance": "_"},
                "depth": 0,
            }
        )

    async def _publish_lifecycle(self, entry: dict[str, Any], state: str, now: int) -> None:
        await self.broker.publish(
            {
                "type": "publish",
                "channel": f"plugins:{entry['id']}:lifecycle",
                "value": {"state": state, "conn": entry["conn"]},
                "ts": now,
                "origin": {"plugin": "kernel", "instance": "_"},
                "depth": 0,
            }
        )

File: internal/broker/registry.go
package broker

import (
	"encoding/json"
	"sync"
	"time"

	"viewer/sdk/go/protocol"
)

type RegistryEntry struct {
	ID          string            `json:"id"`
	InstanceID  *string           `json:"instance_id"`
	Manifest    protocol.Manifest `json:"manifest"`
	Managed     bool              `json:"managed"`
	Conn        string            `json:"conn"`
	ConnectedAt int64             `json:"connected_at"`
}

type Registry struct {
	mu      sync.Mutex
	broker  *Broker
	entries map[string]RegistryEntry
	order   []string
}

func NewRegistry(b *Broker) *Registry {
	return &Registry{broker: b, entries: make(map[string]RegistryEntry)}
}

func (r *Registry) Register(hello protocol.Hello) RegistryEntry {
	r.mu.Lock()
	defer r.mu.Unlock()
	now := time.Now().UnixMilli()
	entry := RegistryEntry{
		ID: hello.Manifest.ID, InstanceID: hello.InstanceID, Manifest: hello.Manifest,
		Managed: hello.Managed, Conn: hello.Conn, ConnectedAt: now,
	}
	if _, exists := r.entries[hello.Conn]; !exists {
		r.order = append(r.order, hello.Conn)
	}
	r.entries[hello.Conn] = entry
	r.publishListLocked(now)
	r.publishLifecycleLocked(entry, "activated", now)
	return entry
}

func (r *Registry) Deregister(conn string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	entry, exists := r.entries[conn]
	if !exists {
		return
	}
	delete(r.entries, conn)
	for i, item := range r.order {
		if item == conn {
			r.order = append(r.order[:i], r.order[i+1:]...)
			break
		}
	}
	now := time.Now().UnixMilli()
	r.publishListLocked(now)
	r.publishLifecycleLocked(entry, "deactivated", now)
}

func (r *Registry) Entries() []RegistryEntry {
	r.mu.Lock()
	defer r.mu.Unlock()
	return r.entriesLocked()
}

func (r *Registry) entriesLocked() []RegistryEntry {
	entries := make([]RegistryEntry, 0, len(r.entries))
	for _, conn := range r.order {
		entries = append(entries, r.entries[conn])
	}
	return entries
}

func (r *Registry) publishListLocked(now int64) {
	value, _ := json.Marshal(r.entriesLocked())
	r.broker.Publish(protocol.Delivery{
		Type: "set", Channel: "plugins:_:list", Value: value, Depth: 0, TS: now,
		Origin: protocol.Origin{Plugin: protocol.KernelPluginID, Instance: protocol.DefaultInstanceID},
	}, "")
}

func (r *Registry) publishLifecycleLocked(entry RegistryEntry, state string, now int64) {
	value, _ := json.Marshal(map[string]any{"state": state, "conn": entry.Conn})
	r.broker.Publish(protocol.Delivery{
		Type: "publish", Channel: "plugins:" + entry.ID + ":lifecycle", Value: value,
		Depth: 0, TS: now,
		Origin: protocol.Origin{Plugin: protocol.KernelPluginID, Instance: protocol.DefaultInstanceID},
	}, "")
}
